rust/opendp-ffi/src/any.rs

```diff
@@ -0,0 +1,593 @@
+        Some(Glue::new(|self_: &Self, other: &Self| {
+            // The first downcast will always succeed, so equality check is all that's necessary.
+            self_.value.downcast_ref::<T>() == other.value.downcast_ref::<T>()
+        }))
+    }
+}
+
+impl<CLONE: Bool, PARTIALEQ: Bool> Downcast for AnyBoxBase<CLONE, PARTIALEQ> {
+    fn downcast<T: 'static>(self) -> Fallible<T> {
+        self.value.downcast().map_err(|_| err!(FailedCast, "Failed downcast of AnyBox to {}", any::type_name::<T>())).map(|x| *x)
+    }
+    fn downcast_ref<T: 'static>(&self) -> Fallible<&T> {
+        self.value.downcast_ref().ok_or_else(|| err!(FailedCast, "Failed downcast_ref of AnyBox to {}", any::type_name::<T>()))
+    }
+}
+
+impl<PARTIALEQ: Bool> Clone for AnyBoxBase<True, PARTIALEQ> {
+    fn clone(&self) -> Self {
+        (self.clone_glue.as_ref().unwrap_assert("No clone_glue for AnyBox"))(&self)
+    }
+}
+
+impl<CLONE: Bool> PartialEq for AnyBoxBase<CLONE, True> {
+    fn eq(&self, other: &Self) -> bool {
+        (self.eq_glue.as_ref().unwrap_assert("No eq_glue for AnyBox"))(self, other)
+    }
+}
+
+/// An AnyBox not implementing optional traits.
+pub type AnyBox = AnyBoxBase<False, False>;
+
+impl AnyBox {
+    pub fn new<T: 'static>(value: T) -> Self {
+        Self::new_base(value, None, None)
+    }
+}
+
+/// An AnyBox implementing Clone.
+pub type AnyBoxClone = AnyBoxBase<True, False>;
+
+impl AnyBoxClone {
+    pub fn new_clone<T: 'static + Clone>(value: T) -> Self {
+        Self::new_base(value, Self::make_clone_glue::<T>(), None)
+    }
+}
+
+/// An AnyBox implementing PartialEq.
+pub type AnyBoxPartialEq = AnyBoxBase<False, True>;
+
+impl AnyBoxPartialEq {
+    pub fn new_partial_eq<T: 'static + PartialEq>(value: T) -> Self {
+        Self::new_base(value, None, Self::make_eq_glue::<T>())
+    }
+}
+
+/// An AnyBox implementing Clone + PartialEq.
+pub type AnyBoxClonePartialEq = AnyBoxBase<True, True>;
+
+impl AnyBoxClonePartialEq {
+    pub fn new_clone_partial_eq<T: 'static + Clone + PartialEq>(value: T) -> Self {
+        Self::new_base(value, Self::make_clone_glue::<T>(), Self::make_eq_glue::<T>())
+    }
+}
+
+/// A struct that can wrap any object.
+pub struct AnyObject {
+    pub type_: Type,
+    value: AnyBox,
+}
+
+impl AnyObject {
+    pub fn new<T: 'static>(value: T) -> Self {
+        Self { type_: Type::of::<T>(), value: AnyBox::new(value) }
+    }
+
+    #[cfg(test)]
+    pub fn new_raw<T: 'static>(value: T) -> *mut Self {
+        crate::util::into_raw(Self::new(value))
+    }
+}
+
+impl Downcast for AnyObject {
+    fn downcast<T: 'static>(self) -> Fallible<T> {
+        self.value.downcast()
+    }
+    fn downcast_ref<T: 'static>(&self) -> Fallible<&T> {
+        self.value.downcast_ref()
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct AnyDomain {
+    pub carrier_type: Type,
+    domain: AnyBoxClonePartialEq,
+    member_glue: Glue<fn(&Self, &<Self as Domain>::Carrier) -> bool>,
```

Comment:
The signature here is clever. &<Self as Domain>::Carrier is just AnyObject. You can check if AnyObject is a member of AnyDomain. AnyDomain may capture all the richness of any impl of Domain.